pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -904,7 +909,14 @@ func (c *Controller) WorkloadInstanceHandler(si *model.WorkloadInstance, event m
 	case model.EventDelete:
 		delete(c.workloadInstancesByIP, si.Endpoint.Address)
 	default: // add or update
+		// Check to see if the workload entry changed. If it did, clear the old entry
+		k := si.Name + "~" + si.Namespace
+		existing := c.workloadInstancesIPsByName[k]
+		if existing != si.Endpoint.Address {
```

Comment:
nit: I think in add case, delete will always be executed which is redundant may be only do this for update case?.